test/e2e/storage/testsuites/base.go

```diff
@@ -465,3 +480,157 @@ func StartPodLogs(f *framework.Framework) func() {
 
 	return cancel
 }
+
+func getVolumeOpsFromMetricsForPlugin(ms metrics.ControllerManagerMetrics, pluginName string) opCounts {
+	totOps := opCounts{}
+
+	for method, samples := range ms {
+		switch method {
+		case "storage_operation_status_count":
+			for _, sample := range samples {
+				plugin := string(sample.Metric["volume_plugin"])
+				if pluginName != plugin {
+					continue
+				}
+				opName := string(sample.Metric["operation_name"])
+				if opName == "verify_controller_attached_volume" {
+					// We ignore verify_controller_attached_volume because it does not call into
+					// the plugin. It only watches Node API and updates Actual State of World cache
+					continue
+				}
+				totOps[opName] = totOps[opName] + int64(sample.Value)
+			}
+		}
+	}
+	return totOps
+}
+
+func getVolumeOpsFromKubeletMetricsForPlugin(ms metrics.KubeletMetrics, pluginName string) opCounts {
+	totOps := opCounts{}
+
+	for method, samples := range ms {
+		switch method {
+		case "storage_operation_status_count":
+			for _, sample := range samples {
+				plugin := string(sample.Metric["volume_plugin"])
+				if pluginName != plugin {
+					continue
+				}
+				opName := string(sample.Metric["operation_name"])
+				if opName == "verify_controller_attached_volume" {
+					// We ignore verify_controller_attached_volume because it does not call into
+					// the plugin. It only watches Node API and updates Actual State of World cache
+					continue
+				}
+				totOps[opName] = totOps[opName] + int64(sample.Value)
+			}
+		}
+	}
+	return totOps
+}
+
+func getVolumeOpCounts(c clientset.Interface, pluginName string) opCounts {
+	metricsGrabber, err := metrics.NewMetricsGrabber(c, nil, true, false, true, false, false)
+
+	if err != nil {
+		framework.Failf("Error creating metrics grabber : %v", err)
+	}
+
+	if !metricsGrabber.HasRegisteredMaster() {
+		framework.Skipf("Environment does not support getting controller-manager metrics - skipping")
+	}
+
+	controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
+	framework.ExpectNoError(err, "Error getting c-m metrics : %v", err)
+	totOps := getVolumeOpsFromMetricsForPlugin(controllerMetrics, pluginName)
+
+	nodes, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	framework.ExpectNoError(err, "Error listing nodes: %v", err)
+	for _, node := range nodes.Items {
```

Comment:
Is the intention that this validation will be checked by default when migration is on by default?

Iterating over all nodes may not work so well for scale testing clusters that are 1000s of nodes.

Reply:
Good catch! I believe most tests specify a "ClientNodeName" which we can use to only query the specific node the volume/pod is scheduled on. However for tests that don't specify what do we want the behavior to be?

Maybe we check how many nodes there are and above a certain threshhold we just decide to not check node metrics? That seems pretty hacky to me, however it still achieves the goal of validating migration functionality when we run the tests on small-ish clusters

Reply:
I've made a 10 node threshhold. However, I've noticed that only certain drivers set a nodeName (local, hostpath, iscsi) while the other drivers do not. I wonder if it is maybe worth having each of the drivers set a nodeName for their tests (which is kind of weird for multi-node tests). However, this seems like an issue with the test infra itself as I believe `nodeName` should probably be a per-test level field, not a per-driver field.

Reply:
clientNodeName essentially bypasses the scheduler and forces the pod to go to a specific node. It should only be used for hostpath.  I'm not sure why local needs it since local supports topology cc @cofyc 

Anyway, we should not use node name, especially as we add more topology and multi-node, multi-writer tests in the future.

Reply:
In that case there is no real generic way (that I can think of) to scrape the metrics from the "correct" node that the test is running on and I'm thinking custom logic for each test doesn't seem great. It would be nice if the operation metric also emitted the node that it was done on, then we could pull which node it got attached to from the controller metric but that increases the carnality of the metric and is only useful for this specific functionality.

Maybe we only look at the controller metrics unless the cluster is "small enough"? That is far from ideal, but it gives us some signal to whether the feature is working or not.. In the near future we were thinking of "disabling" the entire in-tree plugin if the migration feature for that plugin is on anyway, so that combined with some negative metrics checking, combined with tests passing, is a pretty strong signal. Even if the metrics infra part is not the strongest signal on its own